searchcore/src/vespa/searchcore/proton/server/bucketmovejobv2.cpp

```diff
@@ -0,0 +1,368 @@
+    ~IncOnDestruct() {
+        _count.fetch_add(1, std::memory_order_relaxed);
+    }
+private:
+    std::atomic<size_t> & _count;
+};
+
+}
+
+void
+BucketMoveJobV2::prepareMove(BucketMoverSP mover, std::vector<MoveKey> keys, IDestructorCallbackSP onDone)
+{
+    IncOnDestruct countGuard(_executedCount);
+    if (_stopped.load(std::memory_order_relaxed)) return;
+    auto moveOps = mover->createMoveOperations(keys);
+    _master.execute(makeLambdaTask([this, mover=std::move(mover), moveOps=std::move(moveOps), onDone=std::move(onDone)]() mutable {
+        completeMove(std::move(mover), std::move(moveOps), std::move(onDone));
+    }));
+}
+
+void
+BucketMoveJobV2::completeMove(BucketMoverSP mover, std::vector<GuardedMoveOp> ops, IDestructorCallbackSP onDone) {
+    mover->moveDocuments(std::move(ops), std::move(onDone));
+    if (mover->bucketDone() && mover->inSync()) {
+        _modifiedHandler.notifyBucketModified(mover->getBucket());
+    }
+}
+
+void
+BucketMoveJobV2::cancelMovesForBucket(BucketId bucket) {
+    for (auto itr = _movers.begin(); itr != _movers.end(); itr++) {
+        if (bucket == (*itr)->getBucket()) {
+            _movers.erase(itr);
+            backFillMovers();
+            return;
+        }
+    }
+}
+
+void
+BucketMoveJobV2::considerBucket(const bucketdb::Guard & guard, BucketId bucket)
+{
+    ScanIterator itr(guard, bucket);
+    auto [mustMove, wantReady] = needMove(itr);
+    if (mustMove) {
+        _buckets2Move[bucket] = wantReady;
+    } else {
+        _buckets2Move.erase(bucket);
+        cancelMovesForBucket(bucket);
+    }
+    backFillMovers();
+    considerRun();
+}
+
+void
+BucketMoveJobV2::notifyCreateBucket(const bucketdb::Guard & guard, const BucketId &bucket)
+{
+    considerBucket(guard, bucket);
+}
+
+BucketMoveJobV2::BucketSet
+BucketMoveJobV2::computeBuckets2Move()
+{
+    BucketMoveJobV2::BucketSet toMove;
+    for (ScanIterator itr(_ready.meta_store()->getBucketDB().takeGuard(), BucketId()); itr.valid(); ++itr) {
+        auto [mustMove, wantReady] = needMove(itr);
+        if (mustMove) {
+            toMove[itr.getBucket()] = wantReady;
+        }
+    }
+    return toMove;
+}
+
+std::shared_ptr<BucketMover>
+BucketMoveJobV2::createMover(BucketId bucket, bool wantReady) {
+    const MaintenanceDocumentSubDB &source(wantReady ? _notReady : _ready);
+    const MaintenanceDocumentSubDB &target(wantReady ? _ready : _notReady);
+    LOG(debug, "checkBucket(): mover.setupForBucket(%s, source:%u, target:%u)",
+        bucket.toString().c_str(), source.sub_db_id(), target.sub_db_id());
+    return std::make_shared<BucketMover>(bucket, &source, target.sub_db_id(), _moveHandler);
+}
+
+std::shared_ptr<BucketMover>
+BucketMoveJobV2::greedyCreateMover() {
+    if ( ! _buckets2Move.empty()) {
+        auto next = _buckets2Move.begin();
+        auto mover = createMover(next->first, next->second);
+        _buckets2Move.erase(next->first);
```

Comment:
Consider replacing `next->first` with just `next`, to avoid another lookup in the map.

Reply:
Fixed

Reply:
Where ?